test/test_foreach.py

```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
-            if (dtype is torch.float16 or dtype is torch.bfloat16) and TEST_WITH_ROCM:
-                self.assertEqual(tensors1, expected, atol=1.e-3, rtol=self.dtype_precisions[dtype][0])
+    def _inplace_binary_test(self, dtype, inplace, inplace_ref, inputs, is_fastpath, *, alpha=None):
+        requires_cast = self._requires_cast(inplace, dtype, is_fastpath, inputs)
+        ref_inputs = [[t.clone().detach() for t in inputs[0]], inputs[1]]
+        if requires_cast:
+            ref_inputs = [[t.to(torch.float32) for t in ref_inputs[0]], ref_inputs[1]]
+        try:
+            inplace(inputs, self.is_cuda, is_fastpath)
+        except RuntimeError as e:
+            with self.assertRaisesRegex(type(e), re.escape(str(e))):
+                inplace_ref(ref_inputs)
+        else:
+            inplace_ref(ref_inputs)
+            expected = ref_inputs[0]
+            if requires_cast:
+                expected = [t.to(dtype) for t in ref_inputs[0]]
+            self.assertEqual(expected, inputs[0])
+        if alpha is not None:
+            kwargs = {'alpha': alpha}
+            ref_inputs = inputs
+            if requires_cast:
+                ref_inputs = [[t.to(torch.float32) for t in tensors] for tensors in inputs]
+            try:
+                inplace(inputs, self.is_cuda, is_fastpath, **kwargs)
+            except RuntimeError as e:
+                with self.assertRaisesRegex(type(e), re.escape(str(e))):
+                    inplace_ref(ref_inputs, **kwargs)
             else:
-                self.assertEqual(tensors1, expected)
+                inplace_ref(ref_inputs, **kwargs)
+                expected = ref_inputs[0]
+                if requires_cast:
+                    expected = [t.to(dtype) for t in expected]
+                if dtype in (torch.float16, torch.bfloat16) and TEST_WITH_ROCM:
+                    self.assertEqual(expected, inputs[0], atol=1.e-3, rtol=self.dtype_precisions[dtype][0])
+                else:
+                    self.assertEqual(expected, inputs[0])
+
+    def _test_binary_op_tensorlists(self, device, dtype, opinfo, N, is_fastpath, disable_fastpath):
+        n_expected_cudaLaunchKernels = N if disable_fastpath else 1
+        op, ref, inplace_op, inplace_ref = self._get_funcs(opinfo, n_expected_cudaLaunchKernels)
+        inputs = [
+            opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath),
+            opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath),
+        ]
+        self._regular_binary_test(dtype, op, ref, inputs, is_fastpath)
+        self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath)
+        if opinfo.supports_alpha_param:
+            alpha = 3 if dtype in torch.testing.get_all_int_dtypes() else 3.14
+            self._regular_binary_test(dtype, op, ref, inputs, is_fastpath, alpha=alpha)
+            self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath, alpha=alpha)
+
+    @skipMeta
```

Comment:
Should this be skipped for ROCm too?

Reply:
Yes. it's failing because
- this test compares the exception if raised
- foreach with meta tensors raises
- but regular with meta tensors not raise

```
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_add_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_add_meta_float64 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_div_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_div_meta_float64 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_mul_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_mul_meta_float64 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_sub_meta_float32 - AssertionError: NotImplementedError not raised
FAILED test/test_foreach.py::TestForeachMETA::test_binary_op_tensorlists_fastpath__foreach_sub_meta_float64 - AssertionError: NotImplementedError not raised
```